### ISSUE TITLE
Initializing weights

### ISSUE THREAD

**Issue by stefanpeidli** [stefanpeidli/GoNet#9]
**Why is this an important matter?:**
From C. Bishop Neuronal Networks for Pattern Recognition, pages: 260ff
>  We  have 
already  seen that  optimization  algorithms  which  proceed  by a steady  monotonic 
reduction  in  the  error  function  can  become  stuck  in  local  minima.  A  suitable 
choice  of  initial  weights  is therefore  potentially  important  in  allowing  the  train-
ing  algorithm  to  produce  a  good  set  of  weights,  and  in  addition  may  lead  to 
improvements  in the speed  of training.  Even stochastic algorithms such  as gradi-
ent  descent,  which  have the  possibility  of escaping  from  local  minima,  can  show 
strong sensitivity  to  the initial  conditions.

On a very practical level, I have already encountered these problems when setting up weights for the policy net.

**So how should we choose them?:**
> The  majority  of  initialization  procedures  in  current  use  involve  setting  the 
weights  to  randomly  chosen  small  values.  Random  values  are  used  in  order  to 
avoid  problems  due  to  symmetries  in  the  network.  The  initial  weight  values  are 
chosen  to  be  small  so  that  sigmoidal  activation  functions  are  not  driven  into 
the  saturation  regions  where  g'(a)  is  very  small  (which  would  lead  to  small 
grad(E),  and  consequently  a  very  flat  error  surface).  If  the  weights  are  too  small, 
however,  all  of  the  sigmoidal  activation  functions  will  be  approximately  linear, 
which  can  again  lead  to  slow  training.  This  suggests  that  the  summed  inputs 
to  the  sigmoidal  functions  should  be  of  order  unity.  A  random  initialization  of 
the weights  requires that  some choice  be  made  for  the distribution  function  from 
which the weights  are generated.

Specifications on how to derive a distribution for the initial weights can also be found in that chapter (p.261-262).

**In particular there is also a neat idea about having an arrangement of NNs:**

> Since  a  particular  training  run  is  sensitive  to  the  initial  conditions  for  the 
weights,  it  is  common  practice  to  train  a  particular  network  many  times  using 
different  weight  initializations.  This  leads  to  a  set  of  different  networks  whose 
generalization  performance  can  be compared  by making use of independent  data. 
In  this case it is possible to  keep the  best network  and simply discard  the  remain-
der.  However,  improved  prediction  capability  can  often  be  achieved  by  forming 
a  committee  of networks  from  amongst  the  better  ones  found  during  the  training 
process [...].

When comparing NNs with same arquitecture and different weights, we should keep the following rather theoretical feature called _Weight-space  symmetries_ in mind, i.e. (Bishop, p. 133):

> Consider  a  two-layer  network  having  M  hidden  units,  with  'tanh'  activation 
functions  given  by  (4.11) [(tanh)] ,  and  full  connectivity  in  both  layers.  If  we change  the 
sign  of  all  of  the  weights  and  the  bias  feeding  into  a  particular  hidden  unit, 
then,  for  a  given  input  pattern,  the  sign  of  the  activation  of  the  hidden  unit 
will  be  reversed,  since  (4.11)  is  an  odd  function.  This  can  be  compensated  by 
changing  the  sign  of  all  of  the  weights  leading  out  of  that  hidden  unit.  Thus, 
by  changing  the  signs  of  a  particular  group  of  weights  (and  a  bias),  the  input-
output  mapping  function  represented  by  the  network  is  unchanged,  and  so  we 
have  found  two  different  weight  vectors  which  give  rise  to  the  same  mapping 
function.  For  M  hidden  units,  there  will  be  M  such  'sign-flip'  symmetries,  and 
thus  any  given  weight  vector  will  be  one  of  a  set  2^M equivalent  weight  vectors 
(Chen  et  al,  1993).

**Comment by stefanpeidli** [comment]
Choosing the weights like Heining from a normal distribution with mean zero and sigma=1/sqrt(layersize) significantly reduces the K-L divergence. **BUT** now the learning rate has to be chosen far smaller, due to scaling. (0.01->0.0001, otherwise Gradient Descent increases the error, though we could improve Grad Descent here by making it a little bit more clever...)

TODO: Show that inputs are reasonably scaled. Problem might be that they are not. We know that norm(inputvector)~progression in the game. We can expect a bad learning behaviour for sparse and dense boards. Idea: Scale the learning rate with game progression ~ number of stones on the board.

**Comment by stefanpeidli** [comment]
Analysing the Dan_10 games, we observe the following Expectation Value and Standard Deviation of input boards (non-scaled, i.e. -1=white, 0=empty, 1=black):

![image](https://user-images.githubusercontent.com/24811160/33547197-9d1339ca-d8e3-11e7-93e9-93ca6487d1cb.png)

The Expectation Value should be near zero and the Standard deviation near one.
Results:

- We see that the non-scaled inputs are in our sense well-behaved (i.e. reasonably scaled, statistically "good" for our vanilla weight-initialization method).
- Also we observe that the Standard Deviation on near the boundary, particularly in the corners, is slightly (up to 50%) less than on the center of the board. Maybe this can be used when initializing the weights of the first layer in order to further improve learning.

**Comment by stefanpeidli** [comment]
Indeed testing also reveals that the `(B, 0, W) = (-1.35, 0.45, 1.05)` scaling of the inputs yields about the same results:

![image](https://user-images.githubusercontent.com/24811160/33548194-b6981692-d8e6-11e7-88c2-4025423baa96.png)

So we can close this issue and actept the initialization method of 
`W(Layer i)~N(0,1/sqrt(# nodes in Layer i))`.

**Comment by stefanpeidli** [comment]
RELU uses He Initialization.
See [this paper](https://arxiv.org/pdf/1704.08863.pdf).
